x/milestone/abci/abci.go

```diff
@@ -0,0 +1,173 @@
+package abci
+
+import (
+	"fmt"
+	"math/big"
+	"sort"
+	"time"
+
+	"cosmossdk.io/log"
+
+	"github.com/0xPolygon/heimdall-v2/helper"
+	"github.com/0xPolygon/heimdall-v2/sidetxs"
+	"github.com/0xPolygon/heimdall-v2/x/milestone/keeper"
+	"github.com/0xPolygon/heimdall-v2/x/milestone/types"
+	stakeTypes "github.com/0xPolygon/heimdall-v2/x/stake/types"
+	abciTypes "github.com/cometbft/cometbft/abci/types"
+	cmtTypes "github.com/cometbft/cometbft/proto/tendermint/types"
+	"github.com/cosmos/cosmos-sdk/codec/address"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+var pendingMilestoneProposition *sidetxs.MilestoneProposition
+
+func GenMilestoneProposition(ctx sdk.Context, milestoneKeeper *keeper.Keeper, contractCaller helper.IContractCaller) (*sidetxs.MilestoneProposition, error) {
+	propStartBlock := uint64(0)
+
+	milestone, err := milestoneKeeper.GetLastMilestone(ctx)
+	if err != nil && err != types.ErrNoMilestoneFound {
+		return nil, err
+	}
+
+	pendingMilestone := GetPendingMilestoneProposition()
+
+	logger := ctx.Logger()
+
+	timeSinceLastMilestone := time.Duration(0)
+
+	if milestone != nil {
+		timeSinceLastMilestone = time.Since(time.Unix(int64(milestone.Timestamp), 0))
+	}
+
+	logger.Debug("timeSinceLastMilestone", "timeSinceLastMilestone", timeSinceLastMilestone)
+
+	// TODO: make the timeout limit configurable or use a value that depends on block time
+	if pendingMilestone != nil && milestone != nil && timeSinceLastMilestone < 6*time.Second {
+		propStartBlock = pendingMilestone.BlockNumber + 1
+	} else {
+		if milestone == nil {
+			propStartBlock = 0
+		} else {
+			propStartBlock = milestone.EndBlock + 1
```

Comment:
If we timeout, we will propose the same block that we have proposed already in pendingMilestone no?

I think when we implement the API in bor to lock and return blocks hashes, this logic could lead to perpetual logic.

I was imagining that if we don't reach agreement on milestone in PreBlocker, we set in state some flag like `SkipMilestoneHeight = req.Height+1` and we unlock all blocks in same time to allow bor to reorg.
And we check if our req.Height is equal to `SkipMilestoneHeight` and skip all logic for milestones in every ABCI handler.